### PR TITLE
Add metrics for the total number of application threads started in the JVM

### PR DESCRIPTION
**Pull request by lenin-jaganathan**
- generate a metric for total Threads started since the JVM start time. This is extremely useful to understand how an application is doing on thread creation and also understand the creation patterns. The existing metrics give the current status of threads but nothing about whether the threads are created newly or are existing ones. This metric will aid in such scenarios by taking deltas.

### CODE REVIEW

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmGcMetrics.java [hunk 1]
```diff
@@ -170,17 +170,19 @@ public void handleNotification(Notification notification, Object ref) {
...
+            String gcName = notificationInfo.getGcName();
             String gcCause = notificationInfo.getGcCause();
             String gcAction = notificationInfo.getGcAction();
             GcInfo gcInfo = notificationInfo.getGcInfo();
             long duration = gcInfo.getDuration();
-            if (isConcurrentPhase(gcCause, notificationInfo.getGcName())) {
-                Timer.builder("jvm.gc.concurrent.phase.time").tags(tags).tags("action", gcAction, "cause", gcCause)
+            if (isConcurrentPhase(gcCause, gcName)) {
+                Timer.builder("jvm.gc.concurrent.phase.time").tags(tags)
+                        .tags("action", gcAction, "cause", gcCause, "gc_name", gcName)
```
Comment (jonatan-ivanov) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmGcMetrics.java:180]: This is inconsistent to `gc` in `jvm.gc.pause` and I'm thinking if this should be a separate PR.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java [hunk 2]
```diff
@@ -66,6 +66,11 @@ public void bindTo(MeterRegistry registry) {
                 .description("The current number of live threads including both daemon and non-daemon threads")
                 .baseUnit(BaseUnits.THREADS).register(registry);
 
+        Gauge.builder("jvm.threads.started", threadBean, ThreadMXBean::getTotalStartedThreadCount).tags(tags)
```
Comment (jonatan-ivanov) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java:69]: Is there a daemon/live getter so that this can be consistent to `jvm.threads.daemon`/`jvm.threads.live`?

Comment (jonatan-ivanov) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java:69]: This should be monotonic and I guess what users are mostly interested in is not the total number of threads that has been started since the JVM started but the rate of them. So I think this must be a counter. Since there is no event you can use to increment a "normal" `Counter` but there is a (monothonic) method that gives you the current total, you can use a `FunctionCounter` instead of a `Gauge`, see the [docs](https://micrometer.io/docs/concepts#_function_tracking_counters).

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java [hunk 3]
```diff
@@ -66,6 +66,11 @@ public void bindTo(MeterRegistry registry) {
...
+                .description(
+                        "The total number of threads created and also started since the Java virtual machine started")
```
Comment (jonatan-ivanov) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java:71]:
```suggestion
                        "The total number of threads started since the JVM started")
```

**Review by jonatan-ivanov — CHANGES_REQUESTED** [review]
We should add tests for the changes in `JvmGcMetrics` but as I said I might move those to a separate PR.

**Comment by jonatan-ivanov** [comment]
Could you please rebase this on `main`, that should solve the build failure I hope.

**Comment by jonatan-ivanov** [comment]
I was thinking of Loom and the usefulness of this for virtual thread. Also if it is possible to distinguish between virtual and non-virtual threads.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmGcMetrics.java (hunk 1 above)
Comment (lenin-jaganathan) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmGcMetrics.java:180]: I see. How dumb of me to have contrasting names in consecutive conditional blocks 😟. I have no issues having this part of a separate PR.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java (hunk 2 above)
Comment (lenin-jaganathan) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java:69]: Makes sense. But still worried about Step Registry users like us. But it really makes sense to have this as a Counter.

Comment (lenin-jaganathan) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java:69]: Not sure what this means. Can you explain a bit more on this?

Comment (jonatan-ivanov) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java]: There are two types of threads in Java: daemon and non-daemon. See the previous few lines: `jvm.threads.daemon` and `jvm.threads.live`, see: https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.html

Comment (jonatan-ivanov) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java]: Check step registries, for those this will return delta not cumulative.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java [hunk 4]
```diff
@@ -66,6 +67,11 @@ public void bindTo(MeterRegistry registry) {
                 .description("The current number of live threads including both daemon and non-daemon threads")
                 .baseUnit(BaseUnits.THREADS).register(registry);
 
+        FunctionCounter.builder("jvm.threads.started", threadBean, ThreadMXBean::getTotalStartedThreadCount).tags(tags)
+                .description(
+                        "The total number of application threads started in the JVM")
```
Comment (jonatan-ivanov) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java:72]:
```suggestion
                .description("The total number of application threads started in the JVM")
```

**Comment by jonatan-ivanov** [comment]
The build is failing, could you please run `./gradlew format` to fix it?

**Comment by lenin-jaganathan** [comment]
Yes on it @jonatan-ivanov . Having some trouble locally with switching branches. Will update this

**Review by shakuzen — APPROVED** [review]

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java (hunk 2 above)
Comment (lenin-jaganathan) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jvm/JvmThreadMetrics.java]: Yeah. I am aware of that. But those OS implementations give us the current number of threads active. But this is more about the rate at which threads are getting created. And unfortunately, the native implementation doesn't differentiate between daemon and non-daemon for this method.

**Comment by lenin-jaganathan** [comment]
@shakuzen By any chance will this metric be available in 1.10.x versions?

**Comment by jonatan-ivanov** [comment]
@lenin-jaganathan Since this is a new feature, not really. 1.10.x is a maintenance  branch, we only release patch versions which can contain dependency upgrades and bugfixes but it is rare that they contain new features. This should go out early next week (hopefully on Monday) with `1.11.0-M2` and it should go GA in mid May with `1.11.0`.

**Comment by lenin-jaganathan** [comment]
Thanks.
